// nylas/lib/models/application-details.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var model_1 = __importDefault(require("./model"));
var attributes_1 = __importDefault(require("./attributes"));
var ApplicationDetails = /** @class */ (function (_super) {
    __extends(ApplicationDetails, _super);
    function ApplicationDetails(props) {
        var _this = _super.call(this) || this;
        _this.applicationName = '';
        _this.iconUrl = '';
        _this.redirectUris = [];
        _this.initAttributes(props);
        return _this;
    }
    ApplicationDetails.attributes = {
        applicationName: attributes_1.default.String({
            modelKey: 'applicationName',
            jsonKey: 'application_name',
        }),
        iconUrl: attributes_1.default.String({
            modelKey: 'iconUrl',
            jsonKey: 'icon_url',
        }),
        redirectUris: attributes_1.default.StringList({
            modelKey: 'redirectUris',
            jsonKey: 'redirect_uris',
        }),
    };
    return ApplicationDetails;
}(model_1.default));
exports.default = ApplicationDetails;
