// nylas/lib/models/deltas.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var model_1 = __importDefault(require("./model"));
var attributes_1 = __importDefault(require("./attributes"));
var delta_1 = __importDefault(require("./delta"));
var Deltas = /** @class */ (function (_super) {
    __extends(Deltas, _super);
    function Deltas(connection, props) {
        var _this = _super.call(this) || this;
        _this.cursorStart = '';
        _this.cursorEnd = '';
        _this.deltas = [];
        _this.connection = connection;
        _this.initAttributes(props);
        return _this;
    }
    Deltas.attributes = {
        cursorStart: attributes_1.default.String({
            modelKey: 'cursorStart',
            jsonKey: 'cursor_start',
        }),
        cursorEnd: attributes_1.default.String({
            modelKey: 'cursorEnd',
            jsonKey: 'cursor_end',
        }),
        deltas: attributes_1.default.Collection({
            modelKey: 'deltas',
            itemClass: delta_1.default,
        }),
    };
    return Deltas;
}(model_1.default));
exports.Deltas = Deltas;
