// nylas/lib/models/free-busy.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var model_1 = __importDefault(require("./model"));
var attributes_1 = __importDefault(require("./attributes"));
var FreeBusyCalendar = /** @class */ (function (_super) {
    __extends(FreeBusyCalendar, _super);
    function FreeBusyCalendar(props) {
        var _this = _super.call(this) || this;
        _this.accountId = '';
        _this.calendarIds = [];
        _this.initAttributes(props);
        return _this;
    }
    FreeBusyCalendar.attributes = {
        accountId: attributes_1.default.String({
            modelKey: 'accountId',
            jsonKey: 'account_id',
        }),
        calendarIds: attributes_1.default.StringList({
            modelKey: 'calendarIds',
            jsonKey: 'calendar_ids',
        }),
    };
    return FreeBusyCalendar;
}(model_1.default));
exports.FreeBusyCalendar = FreeBusyCalendar;
var TimeSlotCapacity = /** @class */ (function (_super) {
    __extends(TimeSlotCapacity, _super);
    function TimeSlotCapacity(props) {
        var _this = _super.call(this) || this;
        _this.eventId = '';
        _this.currentCapacity = 0;
        _this.maxCapacity = 0;
        _this.initAttributes(props);
        return _this;
    }
    TimeSlotCapacity.attributes = {
        eventId: attributes_1.default.String({
            modelKey: 'eventId',
            jsonKey: 'event_id',
        }),
        currentCapacity: attributes_1.default.Number({
            modelKey: 'currentCapacity',
            jsonKey: 'current_capacity',
        }),
        maxCapacity: attributes_1.default.Number({
            modelKey: 'maxCapacity',
            jsonKey: 'max_capacity',
        }),
    };
    return TimeSlotCapacity;
}(model_1.default));
exports.TimeSlotCapacity = TimeSlotCapacity;
var TimeSlot = /** @class */ (function (_super) {
    __extends(TimeSlot, _super);
    function TimeSlot(props) {
        var _this = _super.call(this) || this;
        _this.object = 'time_slot';
        _this.status = '';
        _this.startTime = 0;
        _this.endTime = 0;
        _this.initAttributes(props);
        return _this;
    }
    TimeSlot.attributes = {
        object: attributes_1.default.String({
            modelKey: 'object',
        }),
        status: attributes_1.default.String({
            modelKey: 'status',
        }),
        startTime: attributes_1.default.Number({
            modelKey: 'startTime',
            jsonKey: 'start_time',
        }),
        endTime: attributes_1.default.Number({
            modelKey: 'endTime',
            jsonKey: 'end_time',
        }),
        emails: attributes_1.default.StringList({
            modelKey: 'emails',
        }),
        capacity: attributes_1.default.Object({
            modelKey: 'capacity',
            itemClass: TimeSlotCapacity,
        }),
    };
    return TimeSlot;
}(model_1.default));
exports.TimeSlot = TimeSlot;
var FreeBusy = /** @class */ (function (_super) {
    __extends(FreeBusy, _super);
    function FreeBusy(props) {
        var _this = _super.call(this) || this;
        _this.object = 'free_busy';
        _this.email = '';
        _this.timeSlots = [];
        _this.initAttributes(props);
        return _this;
    }
    FreeBusy.attributes = {
        object: attributes_1.default.String({
            modelKey: 'object',
        }),
        email: attributes_1.default.String({
            modelKey: 'email',
        }),
        timeSlots: attributes_1.default.Collection({
            modelKey: 'timeSlots',
            jsonKey: 'time_slots',
            itemClass: TimeSlot,
        }),
    };
    return FreeBusy;
}(model_1.default));
exports.default = FreeBusy;
