// nylas/lib/models/neural-sentiment-analysis.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var attributes_1 = __importDefault(require("./attributes"));
var restful_model_1 = __importDefault(require("./restful-model"));
var NeuralSentimentAnalysis = /** @class */ (function (_super) {
    __extends(NeuralSentimentAnalysis, _super);
    function NeuralSentimentAnalysis(connection, props) {
        var _this = _super.call(this, connection, props) || this;
        _this.accountId = '';
        _this.sentiment = '';
        _this.sentimentScore = 0;
        _this.processedLength = 0;
        _this.text = '';
        _this.initAttributes(props);
        return _this;
    }
    NeuralSentimentAnalysis.collectionName = 'sentiment';
    NeuralSentimentAnalysis.attributes = {
        accountId: attributes_1.default.String({
            modelKey: 'accountId',
            jsonKey: 'account_id',
        }),
        sentiment: attributes_1.default.String({
            modelKey: 'sentiment',
        }),
        sentimentScore: attributes_1.default.Number({
            modelKey: 'sentimentScore',
            jsonKey: 'sentiment_score',
        }),
        processedLength: attributes_1.default.Number({
            modelKey: 'processedLength',
            jsonKey: 'processed_length',
        }),
        text: attributes_1.default.String({
            modelKey: 'text',
        }),
    };
    return NeuralSentimentAnalysis;
}(restful_model_1.default));
exports.default = NeuralSentimentAnalysis;
