// nylas/lib/models/scheduler-time-slot.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var attributes_1 = __importDefault(require("./attributes"));
var model_1 = __importDefault(require("./model"));
var SchedulerTimeSlot = /** @class */ (function (_super) {
    __extends(SchedulerTimeSlot, _super);
    function SchedulerTimeSlot(props) {
        var _this = _super.call(this) || this;
        _this.initAttributes(props);
        return _this;
    }
    SchedulerTimeSlot.attributes = {
        accountId: attributes_1.default.String({
            modelKey: 'accountId',
            jsonKey: 'account_id',
        }),
        calendarId: attributes_1.default.String({
            modelKey: 'calendarId',
            jsonKey: 'calendar_id',
        }),
        hostName: attributes_1.default.String({
            modelKey: 'hostName',
            jsonKey: 'host_name',
        }),
        emails: attributes_1.default.StringList({
            modelKey: 'emails',
        }),
        start: attributes_1.default.DateTime({
            modelKey: 'start',
        }),
        end: attributes_1.default.DateTime({
            modelKey: 'end',
        }),
    };
    return SchedulerTimeSlot;
}(model_1.default));
exports.default = SchedulerTimeSlot;
